zerver/actions/create_user.py

```diff
@@ -562,6 +562,12 @@ def do_create_user(
         prereg_realm.created_user = user_profile
         prereg_realm.save(update_fields=["created_user"])
 
+    if realm_creation:
+        from zerver.lib.onboarding import send_initial_realm_messages
+
+        with override_language(realm.default_language):
+            send_initial_realm_messages(realm)
+
```

Comment:
Makes sense. I assume we'll have a test that would fail if this were ordered the other way.

Reply:
Added a test in the last commit that will take care of the ordering here.

Reply:
I think we need something like this to fix the bug posted [here](https://github.com/zulip/zulip/pull/30142#issuecomment-2207003437):

``` diff
diff --git a/zerver/actions/create_user.py b/zerver/actions/create_user.py
index 4b3ba33be8..40d5a36d9b 100644
--- a/zerver/actions/create_user.py
+++ b/zerver/actions/create_user.py
@@ -318,6 +318,12 @@ def process_new_human_user(
     if prereg_user is not None and prereg_user.referred_by is not None:
         notify_invites_changed(user_profile.realm, changed_invite_referrer=prereg_user.referred_by)
 
+    if realm_creation:
+        from zerver.lib.onboarding import send_initial_realm_messages
+
+        with override_language(realm.default_language):
+            send_initial_realm_messages(realm)
+
     notify_new_user(user_profile)
     # Clear any scheduled invitation emails to prevent them
     # from being sent after the user is created.
@@ -595,12 +601,6 @@ def do_create_user(
         prereg_realm.created_user = user_profile
         prereg_realm.save(update_fields=["created_user"])
 
-    if realm_creation:
-        from zerver.lib.onboarding import send_initial_realm_messages
-
-        with override_language(realm.default_language):
-            send_initial_realm_messages(realm)
-
     if bot_type is None:
         process_new_human_user(
             user_profile,

```

Reply:
I think we need a broader reordering, I'll try to make that work.

Reply:
I think an easier way is to use the `realm_creation` parameter for deciding whether to add the `historical` flag or not.

See 5th commit for the changes.

Test added to verify this case

Reply:
Yeah, that works.